File: 07 Risk Measures, Extreme Values, and Copulas/Copulas.py
import pandas as pd
import numpy as np
from copulas.multivariate import GaussianMultivariate
import seaborn as sns
import matplotlib.pyplot as plt

d = pd.read_csv("stock_1999_2002.csv", index_col=0)
returns = np.diff(d, axis=0) / d.iloc[:-1, :] # Arithmetic return
n_sim = 1000

N_cop_dist = GaussianMultivariate()
N_cop_dist.fit(returns)
# Generate random return samples based on multi-variate normal copula
np.random.seed(4002)
return_sim_N = N_cop_dist.sample(n_sim)

sns.pairplot(pd.concat([returns.assign(label='return'), 
                        return_sim_N.assign(label='sim_return')]), 
             hue='label', diag_kind='kde', palette=['orange', 'blue'])

plt.tight_layout()
plt.savefig("../Picture/Gaussian_Copula.png", dpi=200)

#%%
import statsmodels.distributions.empirical_distribution as edf
from scipy.interpolate import interp1d

def Empirical_QQ_plot(sim_data, returns, col=None):
    n_stocks = len(returns.columns)
    n_days   = len(returns)
    if col is None: col = ["black"] * n_stocks
    
    fig, axes = plt.subplots(ncols=n_stocks, figsize=(10*n_stocks, 10))
    for k in range(n_stocks):
        sim_data_k = sim_data.iloc[:,k]
        returns_k  = returns.iloc[:,k]
        sim_data_k_quantile = edf.ECDF(sim_data_k)(sim_data_k)
        inverted_k_edf = interp1d(sim_data_k_quantile, sim_data_k)
        i = (np.arange(1, n_days + 1) - 0.5) / n_days
        q = inverted_k_edf(i)
        
        b, w = np.linalg.lstsq(np.c_[np.ones(n_days), q], 
        					   np.sort(returns_k), rcond=None)[0]
        axes[k].scatter(q, np.sort(returns_k), color=col[k],
                        facecolor="white", marker="o", s=50)
        axes[k].plot(q, w*q + b, color="blue", linewidth=2)
        axes[k].set_xlabel("Empirical quantiles", fontsize=15)
        axes[k].set_ylabel("Returns quantiles", fontsize=15)
        axes[k].set_title(f"{d.columns.values[k]} Empirical Q-Q Plot", 
                          fontsize=20)
    
    return fig

total_sim = 1e5; col = ["blue", "orange", "green"]

#%%
np.random.seed(4002)
sim_N = N_cop_dist.sample(int(total_sim))
fig = Empirical_QQ_plot(sim_N, returns, col=col)

fig.tight_layout()
fig.savefig("../Picture/Empirical N QQ Plot.png", dpi=200)
#%%
from copulas.univariate import StudentTUnivariate

t_cop_dist = GaussianMultivariate(distribution={
    "HSBC": StudentTUnivariate,
    "CLP": StudentTUnivariate,
    "CK": StudentTUnivariate})

t_cop_dist.fit(returns)
# Generate random return samples based on multi-variate t copula
np.random.seed(4002)
return_sim_t = t_cop_dist.sample(n_sim)

sns.pairplot(pd.concat([returns.assign(label='return'), 
                        return_sim_t.assign(label='sim_return')]), 
             hue='label', diag_kind='kde', palette=['orange', 'blue'])

plt.tight_layout()
plt.savefig("../Picture/t_Copula.png", dpi=200)

#%%
np.random.seed(4002)
sim_t = t_cop_dist.sample(int(total_sim))
fig = Empirical_QQ_plot(sim_t, returns, col=col)

fig.tight_layout()
fig.savefig("../Picture/Empirical t QQ Plot.png", dpi=200)

File: 07 Risk Measures, Extreme Values, and Copulas/Copulas.R
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

d <- read.csv("stock_1999_2002.csv", row.names=1) # read in data file
d <- as.ts(d)
return <- (lag(d) - d)/d
colnames(return) <- paste0(colnames(d), "_Return")
u1 <- return[,"HSBC_Return"]
u2 <- return[,"CLP_Return"]
u3 <- return[,"CK_Return"]
n_sim <- 1000

library(copula)  # Package for copula computation
# Assume a normal-copula with ncol(d)=3
# P2p: array of elements of upper triangular matrix
N.cop <- normalCopula(P2p(cor(d)), dim=ncol(d), dispstr="un")
set.seed(4002)
# Generate random samples u~U(0, 1) based on gaussian copula
u_sim <- rCopula(n_sim, N.cop)
pairs(u_sim, col="blue")
cor(u_sim)
cor(d)

# mvdc:  Multi-Variate Distribution Copula
N_cop_dist <- mvdc(copula=N.cop, margins=rep("norm", ncol(d)),
                   paramMargins=list(list(mean=mean(u1), sd=sd(u1)),
                                     list(mean=mean(u2), sd=sd(u2)),
                                     list(mean=mean(u3), sd=sd(u3))))

# Generate random return samples based on multi-variate normal copula
set.seed(4002)
return_sim_N <- rMvdc(n_sim, N_cop_dist)
pairs(rbind(return, return_sim_N), col=c("orange","blue"))

######################################################################
Empirical_QQ_Plot <- function(sim_data, returns, col=c()){
  n_stocks <- ncol(returns); n_days <- nrow(returns)
  if (length(col) == 0) col <- rep("black", n_stocks)
  
  par(mfrow=c(1, n_stocks))
  for (k in 1:n_stocks){
    sim_data_k <- sim_data[,k]; returns_k <- returns[,k]
    i <- ((1:n_days) - 0.5) / n_days
    q <- quantile(ecdf(sim_data_k), probs=i)
    
    qqplot(q, sort(returns_k), xlab="Empirical quantiles",
           ylab="Returns quantiles", col=col[k],
           main=paste(colnames(returns)[k], "Empirical Q-Q Plot"))
    abline(lsfit(q, sort(returns_k)))
  }
}

total_sim <- 1e4; col <- c("blue", "orange", "green")

######################################################################
set.seed(4002)
sim_N <- rMvdc(total_sim, N_cop_dist)
Empirical_QQ_Plot(sim_N, return, col=col)

######################################################################

# Assume a t-copula  with ncol(d)=3
t.cop <- tCopula(dim=ncol(d), dispstr='un')
m <- pobs(return)
fit <- fitCopula(t.cop, m, "ml")
(rho <- coef(fit)[1:ncol(d)])
(df <- coef(fit)[length(coef(fit))])

# Generate random samples u~U(0, 1) based on the fitted t copula
u_sim <- rCopula(n_sim, tCopula(dim=ncol(d), rho, df=df, dispstr="un"))
pairs(u_sim, col="blue")
cor(u_sim)
cor(d)

t.nu <- function(u){
  z <- u - mean(u)              # Remove mean
  n <- length(z)                # sample size
  s <- sd(z)*sqrt((n-1)/n)      # Population standard deviation
  ku <- sum(z^4)/(n*s^4) - 3    # Excess kurtosis
  6/ku + 4                      # Degrees of freedom
}

t_cop_dist <- mvdc(copula=tCopula(rho, dim=ncol(d), df=df, dispstr="un"),
                   margins=rep("norm", ncol(d)),
                   paramMargins=list(list(mean=mean(u1), sd=sd(u1)),
                                     list(mean=mean(u2), sd=sd(u2)),
                                     list(mean=mean(u3), sd=sd(u3))))

# Generate random return samples based on multi-variate t copula
set.seed(4002)
return_sim_t <- rMvdc(n_sim, t_cop_dist)

pairs(rbind(return, return_sim_t), col=c("orange","blue"))

######################################################################
set.seed(4002)
sim_t <- rMvdc(total_sim, t_cop_dist)
Empirical_QQ_Plot(sim_t, return, col=col)
